### ISSUE TITLE
Provide support for C functional macros

### ISSUE THREAD

**Issue by Slabity** [jethrogb/rust-cexpr#3]
Currently, cexpr cannot produce values that use functional macros to calculate. It would be very useful if cexpr supported them.

**Comment by jethrogb** [comment]
cexpr support is almost trivial, assuming these functional macros evaluate into proper expressions. If this is useful I can implement it. Note however that you must indicate to cexpr whether you'd be parsing a functional macro or a regular macro, because there is no way to distinguish between the two at the token level.

**Comment by Slabity** [comment]
> there is no way to distinguish between the two at the token level.

How so?

**Comment by jethrogb** [comment]
Both of
```
m(a) -1
m (a) -1
```
are
```
[
  Identifier("m"),
  Punctuation("("),
  Identifier("a"),
  Punctuation(")"),
  Punctuation("-"),
  Literal("1")
]
```

**Comment by Slabity** [comment]
Could there be a `Whitespace` token to differentiate the two?

**Comment by jethrogb** [comment]
There could. But you need a lexer that generates those tokens. Clang does not.

**Comment by emilio** [comment]
Hi, @jethrogb, would it be fine to give you a boolean that tells whether the current macro definition is function-like?

Clang has an API for that (`clang_Cursor_isMacroFunctionLike`), and it'd solve some bindgen issues like https://github.com/servo/rust-bindgen/issues/753

**Comment by jethrogb** [comment]
Yes that would work. cexpr just needs to know whether it's parsing a functional or regular macro.

Note: clang_Cursor_isMacroFunctionLike was introduced relatively recently and is e.g. not available on Ubuntu 16.04.

Also notably, clang_Cursor_Evaluate was introduced in the same commit, but I think it's not as powerful as cexpr.

**Comment by emilio** [comment]
Right both are libclang 3.9+. But nicely enough bindgen loads libclang at runtime so can detect which functions are available, so it'd be nice to start using them when available.

**Comment by nicokoch** [comment]
Any update on this? :)

**Comment by jethrogb** [comment]
It was too hot to do any work ;) Next weekend... Or feel free to pick it up yourself before then.

**Comment by Slabity** [comment]
@jethrogb Any update?

**Comment by jethrogb** [comment]
Was going to do this this morning, but got distracted by 0ea1367 and 4fdd26b

**Comment by jethrogb** [comment]
Parsing functional macro declarations is now supported in fe05507.

I supposed you'd also like to evaluate them if they're used as part of a another macro definition? That should not be too hard, does anyone know what the precendence is of the function call operator?

**Comment by emilio** [comment]
Awesome!

Yeah, I think having support to resolve a macro inside other macro is what can make it useful for Bindgen... Otherwise there isn't much utility on it... I think the most usual request is for simple macros that resolve to numeric types, for which there shouldn't be any token-pasting issue I guess? Something like:

```c
#define MY_FLAG(i) (1 << (i))

#define MY_VAR_1 MY_FLAG(1)
#define MY_VAR_2 MY_FLAG(2)
#define MY_VAR_3 MY_FLAG(3)
```

Supporting that would be quite nice IMO.

**Comment by jethrogb** [comment]
I've investigated this further and I don't think it makes sense to not use a mostly-implemented mostly-compliant C preprocessor. The preprocessor is completely separate from the expression evaluation implemented by this crate. The preprocessor would basically take a macro replacement list and fully expand the list of tokens which could then be evaluated by `cexpr`. There are a couple of options:
1. Figure out how to leverage Clang's preprocessor. I looked into this but the C API doesn't seem to have anything. There's clang_Cursor_Evaluate which could do both jobs but it has some limitations, for example it doesn't distinguish `char`s from `int`s. The C++ API might be useful according to [this StackOverflow post](https://stackoverflow.com/questions/39529480/is-there-a-way-to-get-source-code-with-macro-expanded-using-clang-api) but that's an unstable API with not dynamic libraries.
2. Incorporate a stand-alone preprocessor such as [mcpp](http://mcpp.sourceforge.net/).
3. Write one in Rust

**Comment by photoszzt** [comment]
Would this be useful: https://stackoverflow.com/questions/13881506/retrieve-information-about-pre-processor-directives 
https://stackoverflow.com/questions/10113586/how-can-i-parse-macros-in-c-code-using-clang-as-the-parser-and-python-as-the ?

**Comment by jethrogb** [comment]
No. We already do that.

**Comment by photoszzt** [comment]
Directly calling clang -E isn't an option?

**Comment by Slabity** [comment]
Any progress on this? It's still impossible to use the `_IO` family of macros.

I use to be able to hack around this by rebinding them as constants, but this no longer works for some reason.

**Comment by Slabity** [comment]
Pinging @jethrogb - Anything I can do to help fix this?

**Comment by jethrogb** [comment]
@Slabity Yes, the next steps are described in https://github.com/jethrogb/rust-cexpr/issues/3#issuecomment-315645257

**Comment by gnzlbg** [comment]
So an `mpi-sys` crate needs to expose constants that are implemented slightly differently by the different MPI implementations, for example, see `MPI_REQUEST_NULL` in:

* [mpich](https://github.com/pmodels/mpich/blob/45fc41144de975f5bf938c87c036e2231166f269/src/include/mpi.h.in#L93):

```c
 #define MPI_REQUEST_NULL   ((MPI_Request)0x2c000000)
```

* [openmpi](https://github.com/open-mpi/ompi/blob/c79a4c55a24fba10edfbfa2d5a75ac11e950021e/ompi/include/mpi.h.in#L754)

```c
#define MPI_REQUEST_NULL OMPI_PREDEFINED_GLOBAL(MPI_Request, ompi_request_null)
```

it would be great if `rust-bindgen` could import these as `const`s, instead of as `static mut`.

**Comment by pvdrz** [comment]
> I've investigated this further and I think best results would be obtained with a mostly-implemented mostly-compliant C preprocessor. The preprocessor is completely separate from the expression evaluation implemented by this crate. The preprocessor would basically take a macro replacement list and fully expand the list of tokens which could then be evaluated by `cexpr`. There are a couple of options:
> 
>     1. Figure out how to leverage Clang's preprocessor. I looked into this but the C API doesn't seem to have anything. There's clang_Cursor_Evaluate which could do both jobs but it has some limitations, for example it doesn't distinguish `char`s from `int`s. The C++ API might be useful according to [this StackOverflow post](https://stackoverflow.com/questions/39529480/is-there-a-way-to-get-source-code-with-macro-expanded-using-clang-api) but that's an unstable API with no dynamic libraries.
> 
>     2. Incorporate a stand-alone preprocessor such as [mcpp](http://mcpp.sourceforge.net/).
> 
>     3. Write one in Rust.

:wave: I'm writing a C preprocessor from scratch and I'd like to know what kind of API do you have in mind for the expansion. Would you pass the preprocessing directive defining the macro and the tokens where the macro is supposed to be expanded?

**Comment by photoszzt** [comment]
Before writing a new one, what's missing from the mcpp? Is it just hard to use in Rust?

**Comment by pvdrz** [comment]
Well you'd have to write rust bindings for `mcpp` by hand (or using `bindgen` but we would have to see if `bindgen` can generate valid rust bindings for all the items). At the same time the project seems to be unmaintained, the last release was in 2008 and the last update to their sourceforge was 2013. It could be fine as they are C99 compliant but who knows.
